### PR TITLE
Extract spatially local sub-templates from `Template` instances

### PR DESCRIPTION
**Pull request by nelimee**
This is based on #317.

This PR introduces several things to the `Template` hierarchy. First, a few details:
- Remove `pass` from abstract methods bodies
- Introduce a `RectangularTemplate` class that inherits from `Template` and provides a default implementation of `get_midline_plaquettes` for `Template` instances with a rectangular form (e.g. all the templates in `tqec.templates.qubit`).

Now the main 2 changes:
- Introduction of a `GridTemplate` that allows to quickly represent a grid of several identical `RectangularTemplate` instances.
- Introduction of a `get_spatially_distinct_subtemplates` function that computes, from a template, all the spatially-local plaquette arrangements found.

A little bit more on `get_spatially_distinct_subtemplates`. It takes an array obtained from calling `Template.instantiate` on any `Template` instance, a `radius` (using the Manhattan distance), and compute all the plaquette arrangements that are found in the instantiation with the provided `radius`.

For example, a radius of `1` would lead to "sub-template" looking like (`c` being the center of the ball of radius `1`):
```
x x x
x c x
x x x
```
whereas a radius of 3 would lead to
```
x x x x x x x
x x x x x x x
x x x x x x x
x x x c x x x
x x x x x x x
x x x x x x x
x x x x x x x
```

Now with a real example, the following code
```py
from tqec.templates.qubit import QubitTemplate

template = QubitTemplate(k=2)
situation, mapping = template.get_spatially_distinct_subtemplates(1)

print("Full map:")
display_template_from_instantiation(situation)
for i, subtemplate in mapping.items():
    print(f"Subtemplate {i}:")
    display_template_from_instantiation(subtemplate)
```
outputs:
```
Full map:
  1  2  4  6  3  5
  7 12 14 16 13 15
  9 18 21 28 24 33
 11 20 28 21 31 26
  8 17 22 29 25 34
 10 19 30 23 32 27
Subtemplate 1:
  .  .  .
  .  1  5
  .  7  9
Subtemplate 2:
  .  .  .
  1  5  6
  7  9 10
Subtemplate 3:
  .  .  .
  5  6  2
  9 10 11
Subtemplate 4:
  .  .  .
  5  6  5
  9 10  9
// Cut for readability
Subtemplate 16:
  6  5  6
 10  9 10
  9 10  9
Subtemplate 17:
  7  9 10
  8 10  9
  3 13 14
// Cut for readability
Subtemplate 33:
 10 11  .
  9 12  .
 10 11  .
Subtemplate 34:
 10 11  .
  9 12  .
 14  4  .
```
In theory, if we ignore `0` entries in the sub-templates (represented by `.` above), we should recover exactly the template instantiation by superimposing on each entry of the "Full map" printed first the corresponding sub-template.

The goal of implementing such a procedure is to be able to list exhaustively the spatially-local sub-templates that we have to consider in order to compute detectors involving measurements in the center plaquette of such sub-template.
What is interesting in this approach is that if we now use a way larger `Template` such as
```py 
from tqec.templates.qubit import QubitTemplate

template = QubitTemplate(k=100)  # << WAY LARGER
situation, mapping = template.get_spatially_distinct_subtemplates(1)

print("Full map:")
display_template_from_instantiation(situation)
for i, subtemplate in mapping.items():
    print(f"Subtemplate {i}:")
    display_template_from_instantiation(subtemplate)
```
the number of distinct spatially local sub-template is still `34`, and we have in the variable `situation` a description of where these sub-templates are.

One of the issue I still have is about naming: I do not like the naming of `situation` to describe a 2-dimensional array filled with sub-template indices, but I cannot find something better. Do anyone have an idea?

### CODE REVIEW

**Comment by inmzhang** [comment]
Is this output correct? Why is the subtemplate centered at 1 is:
```txt
.  .  .
.  1  5
.  7  9
```
instead of
```txt
.  .  .
.  1  2
.  7  12
```

**Comment by nelimee** [comment]
> Is this output correct? Why is the subtemplate centered at 1 is:
> 
> ```
> .  .  .
> .  1  5
> .  7  9
> ```
> 
> instead of
> 
> ```
> .  .  .
> .  1  2
> .  7  12
> ```

It is correct because the `QubitTemplate.instantiate` method
```py
from tqec.templates.display import display_template

template = QubitTemplate(k=2)
display_template(template)
```
returns
```
  1  5  6  5  6  2
  7  9 10  9 10 11
  8 10  9 10  9 12
  7  9 10  9 10 11
  8 10  9 10  9 12
  3 13 14 13 14  4
```

File: src/tqec/templates/subtemplates.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+            but that is filled with indices from the second part of this
+            tuple.
+        2. a mapping from indices to the corresponding sub-template.
+    """
+    y, x = instantiation.shape
+    vzeros = numpy.zeros((y + 2 * manhattan_radius, manhattan_radius), dtype=numpy.int_)
+    hzeros = numpy.zeros((manhattan_radius, x), dtype=numpy.int_)
+    extended_instantiation = numpy.hstack(
+        (vzeros, numpy.vstack((hzeros, instantiation, hzeros)), vzeros)
+    )
```
[the remark this answers is not in the document]
Comment (nelimee) [reply src/tqec/templates/subtemplates.py]: Right, I was not aware of `numpy.pad`, thanks for letting me know that it exists :+1:

File: src/tqec/templates/subtemplates.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+            (i.e., 0 value in the instantiation of the Template instance) at
+            its center should be ignored. Default to `True`.
+
+    Returns:
+        a tuple containing:
+
+        1. an array that has the exact same shape as `self.instantiate()`
+            but that is filled with indices from the second part of this
+            tuple.
+        2. a mapping from indices to the corresponding sub-template.
```
[the remark this answers is not in the document]
Comment (nelimee) [reply src/tqec/templates/subtemplates.py]: Is the last commit addressing correctly this issue, or is the documentation still lacking?

[the remark this answers is not in the document]
Comment (inmzhang) [reply src/tqec/templates/subtemplates.py]: Great, it's much better now.

**Review by inmzhang — APPROVED** [review]

**Pull request merged**